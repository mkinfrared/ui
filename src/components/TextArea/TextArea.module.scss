@import "../../styles/common.scss";

.TextArea {
  --background: var(--black-light);
  --border-color: var(--grey-light);
  --color: var(--white);

  :where(&) {
    @include theme("light") {
      --background: var(--white);
      --border-color: var(--grey-dark);
      --color: var(--black);
    }

    @include theme("dark") {
      --background: var(--black-light);
      --border-color: var(--grey-light);
      --color: var(--white);
    }
  }

  .label {
    cursor: pointer;
    margin-bottom: 1rem;
    text-transform: capitalize;
  }

  .textArea {
    background-color: var(--background);
    border: 1px solid;
    border-color: var(--border-color);
    border-radius: 4px;
    caret-color: var(--secondary);
    color: var(--color);
    font-size: 1.6rem;
    line-height: 24px;
    outline: none;
    overflow: hidden;
    padding: 5px 10px;
    resize: none;
    width: 100%;

    &[aria-invalid="true"] {
      --border-color: var(--error);

      &:hover {
        border-color: var(--error);
      }
    }

    &::placeholder {
      color: var(--grey);
    }

    &:-webkit-autofill,
    &:-webkit-autofill:active,
    &:-webkit-autofill:hover,
    &:-webkit-autofill:focus {
      background-color: var(--background-color);
      border: none;
      box-shadow: 0 0 0 1000px var(--background-color) inset;
      caret-color: var(--secondary);
      font-size: 100%;
      -webkit-text-fill-color: var(--white);

      // it's not possible to use transparent background color
      // to override chrome autofill styles, so background needs
      // to be set to a certain color, to keep the initial background
      // transparent the transition delay can be used
      transition: background-color 7200s ease-in-out 0s; // 2 hr delay
    }

    &::selection {
      background-color: var(--cyan);
      color: var(--white);
    }

    &:focus-visible {
      outline: none;
    }
  }

  .error {
    color: var(--error);
  }
}
