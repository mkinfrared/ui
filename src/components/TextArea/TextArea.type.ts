import { ChangeEventHandler, Ref } from "react";

export type TextAreaProps = {
  autogrow?: boolean;
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  defaultValue?: string;
  disabled?: boolean;
  /**
   * an error text to be displayed under input field
   * if value is truthy also changes css styles to display error
   */
  error?: string;
  label?: string;
  name?: string;
  onChange?: ChangeEventHandler<HTMLTextAreaElement>;
  placeholder?: string;
  rows?: number;
  textAreaRef?: Ref<HTMLTextAreaElement>;
  value?: string;
};
