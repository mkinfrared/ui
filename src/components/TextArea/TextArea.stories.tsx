import { Meta, Story } from "@storybook/react";

import css from "./Story.module.scss";
import TextArea from "./TextArea";
import { TextAreaProps } from "./TextArea.type";

export default {
  title: "UI/TextArea",
  component: TextArea,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<TextAreaProps> = (args) => (
  <div className={css.Story}>
    <TextArea {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {
  label: "Foobar",
  value:
    "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Architecto culpa dignissimos, iure qui repudiandae voluptate? Laboriosam modi natus nisi quas.",
};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
