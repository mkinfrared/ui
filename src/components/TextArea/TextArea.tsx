import { uniqueId } from "lodash-es";
import { ChangeEventHandler, useMemo, useRef } from "react";

import Heading from "components/Heading";
import Text from "components/Text";
import TextAreaInput from "components/TextAreaInput";
import { classNames, mergeRefs } from "utils";

import css from "./TextArea.module.scss";
import { TextAreaProps } from "./TextArea.type";

const TextArea = ({
  autogrow = true,
  className,
  defaultValue,
  disabled = false,
  error,
  label,
  name,
  onChange,
  placeholder,
  rows = 4,
  textAreaRef,
  value,
}: TextAreaProps) => {
  const initialRows = rows;
  const ref = useRef<HTMLTextAreaElement | null>(null);
  const errorId = useMemo(() => uniqueId("error"), []);

  const handleChange: ChangeEventHandler<HTMLTextAreaElement> = (event) => {
    const { current } = ref;

    if (current && autogrow) {
      current.setAttribute("rows", "2");

      const { lineHeight } = window.getComputedStyle(current);
      const { scrollHeight } = current;
      const lines = Math.floor(scrollHeight / parseInt(lineHeight, 10));
      const rowsAmount = Math.max(lines, initialRows);

      current.setAttribute("rows", `${rowsAmount}`);
    }

    onChange?.(event);
  };

  return (
    <label
      className={classNames(css.TextArea, className)}
      data-testid="TextArea"
    >
      {!!label && (
        <Heading variant="h5" className={css.label}>
          {label}
        </Heading>
      )}
      <TextAreaInput
        aria-invalid={!!error}
        aria-errormessage={errorId}
        className={css.textArea}
        defaultValue={defaultValue}
        disabled={disabled}
        name={name}
        onChange={handleChange}
        placeholder={placeholder}
        rows={initialRows}
        textAreaRef={mergeRefs(ref, textAreaRef)}
        value={value}
      />
      <Text aria-live="polite" className={css.error} id={errorId}>
        {error}
      </Text>
    </label>
  );
};

export default TextArea;
