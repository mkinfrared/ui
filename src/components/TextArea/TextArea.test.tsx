import { fireEvent, render } from "@testing-library/react";

import TextArea from "./TextArea";

describe("<TextArea />", () => {
  const onChange = jest.fn();
  const Component = <TextArea onChange={onChange} />;

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should be defined", () => {
    expect(TextArea).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("TextArea");

    expect(element).toBeDefined();
  });

  it("should render label if it's truthy", () => {
    const { getByTestId } = render(<TextArea label="Marklar" />);
    const element = getByTestId("TextArea");
    const label = element.querySelector(".label");

    expect(label).not.toBeNull();
  });

  it("should change the row attribute on change event", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("TextArea");
    const textArea = element.querySelector(".textArea") as HTMLTextAreaElement;
    const mockGetComputedStyle = jest.spyOn(window, "getComputedStyle");
    const mockScrollHeight = jest.spyOn(textArea, "scrollHeight", "get");

    mockGetComputedStyle.mockImplementation(() => ({ lineHeight: 100 } as any));

    mockScrollHeight.mockImplementation(() => 1000);

    fireEvent.change(textArea, { target: { value: "S" } });

    expect(textArea.rows).toBe(10);
  });

  it("should call 'onChange' from props on change event", () => {
    const { getByTestId } = render(
      <TextArea onChange={onChange} autogrow={false} />,
    );

    const element = getByTestId("TextArea");
    const textArea = element.querySelector(".textArea") as HTMLTextAreaElement;
    const mockGetComputedStyle = jest.spyOn(window, "getComputedStyle");

    mockGetComputedStyle.mockImplementation(() => ({ lineHeight: 100 } as any));

    fireEvent.change(textArea, { target: { value: "S" } });

    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
