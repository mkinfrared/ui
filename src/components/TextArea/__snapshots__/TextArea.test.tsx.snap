// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should match the snapshot 1`] = `
<div>
  <label
    class="TextArea"
    data-testid="TextArea"
  >
    <textarea
      aria-errormessage="error1"
      aria-invalid="false"
      class="textArea"
      data-testid="TextAreaInput"
      rows="4"
    />
    <p
      aria-live="polite"
      class="Text error"
      data-testid="Text"
      id="error1"
    />
  </label>
</div>
`;
